### ISSUE TITLE
Receiving interface rx values as both rx and tx

### ISSUE THREAD

**Issue by dimrozakis** [perfwatcher/collectm#7]
I'm noticing something strange. It seems that for each network interface, the rx and tx metrics sample the same data source. I tried downloading a large file from the monitored server, and both rx and tx dataseries displayed exactly the same spikes. Could this be a bug in collectm's collectdCompat plugin? It seems like the interface metrics (if_octets, if_packets) send the rx values properly as rx, but incorrectly as tx as well.

**Comment by ymettier** [comment]
Hello,

Could you test Collectm-1.5.0 (just released) ?
You should not see any difference because nothing changed in the code relative to interfaces.
It's only to be sure that we work with the same versions.

Then, could you replace `collectdCompat.js` from Collectd-1.5.0 with the one from test_issue_7 branch ?
https://github.com/perfwatcher/collectm/blob/test_issue_7/src/plugins/collectdCompat.js

You can see here what I changed : 
https://github.com/perfwatcher/collectm/commit/1e080b1d09253377e6ab09d3dfe12fe725b5639c#diff-9fdbdec5812e5eb7a0143dbccc985a18

Thanks in advance for your feedback.
Yves

**Comment by dimrozakis** [comment]
Hi Yves,

I installed 1.5.0, no change as expected. I replaced collectdCompat.js with the one from test_issue_7 branch, still no change. Btw, I happened to tried the exact same patch yesterday before opening the issue. It didn't work then, it still doesn't.

**Comment by ymettier** [comment]
Right now, I have no server with several interfaces. So I cannot reproduce the problem.

Here is the algo we run, if you can give a try and find what is going wrong : 

1/ get the list of interfaces
```
C:\> typeperf -qx "Network Interface"
```

2/ then, for each line, run typeperf with that metric. Example : 
```
C:\> typeperf "\Network Interface(your interface here)\Bytes Received/sec"
```

If you can do that, could you check if the problem comes from typeperf, from how Collectm uses typeperf, or a bug in Collectm code ?

Note : you can also use the log this way : 
at the beginning of collectdCompat.js, add this line (useful for `util.inspect()`):
```
var util = require('util');
```
Then, when you want, use `logger.log('info', 'DEBUG : your message')`. Example :
```
function get_interface() { // {{{
    perfmon.list('Network Interface', function(err, datas) {
        var newcounters = datas.counters.sort();
        if((newcounters.length!=known_interfaces.length)
                || ! (newcounters.every(function(v,i) { return (v === known_interfaces[i]); } ))) {
            known_interfaces = newcounters;
			logger.log('info', 'DEBUG : Found interfaces : '+util.inspect(known_interfaces, { showHidden: true, depth: null}));

            perfmon(known_interfaces, function(err, data) {
```

Then check the log file in logs directory.

Note for me : if there is a bug in Collectm code, also check `get_disk()` that has a similar code to `get_interface()`.

Regards,
Yves

**Comment by dimrozakis** [comment]
Hey Yves

Good news! I fixed the bug. It turns out it was a bug in collectdout module. Please see https://github.com/perfwatcher/collectm/commit/243e74599a227ffdfad501d830796219ba66d9f0. In the process of tracking down the bug, I also simplified `get_interface` in https://github.com/perfwatcher/collectm/commit/6a40b4b4d3716be0de7fc4651f0243a0551859a6. Could you please review the changes and let me know if you'd like me to open a pull request containing one or both commits?

**Comment by dimrozakis** [comment]
Would you rather I open a PR in https://github.com/astro/node-collectdout?

**Comment by ymettier** [comment]
Hello,

Just checked commit 243e74599a227ffdfad501d830796219ba66d9f0. It looks OK for me. Nice job !
Yes, please open a PR in  https://github.com/astro/node-collectdout and we will then merge the new version into collectm.
@feraudet, could you please follow this PR ? I will follow it to merge it into a Collectd-1.5.1 version asap then.

@dimrozakis, about the simplification of `get_interface()`, I had a very quick look and I don't like the `[value, 0]` because it would insert 0 as a value. If I'm wrong, please forget this comment. I will take more time tomorrow to read it with more attention.

Regards,
Yves

**Comment by dimrozakis** [comment]
I'll open a PR on https://github.com/astro/node-collectdout linking it to this issue. As far as I can tell, this bug affects all plugins with multiple counter values.

On the simplification of `get_interface()`:

While debugging this issue, I added several log entries and noticed that even though I had configured an interval of 10 seconds, interface data was actually being collected once per second. Given how if_octets, if_packets and if_errors are counters, and judging by the code in collectdout's `addCounter`, i think that `addCounter` adds the submitted values (increments) to the stored values (every 1 sec) and then flushes them according to the configured interval (10 secs in my case). So `[value, 0]` just adds 0 to the second counter (tx in this case).

If you don't like this approach, then the commit in your test_issue_7 branch should perhaps be used instead, cause I fear that the existing implementation (in your master branch) will submit most of the values to be added twice.

Also, if you chose not to use my second commit, I think that https://github.com/perfwatcher/collectm/blob/master/src/plugins/collectdCompat.js#L223 and https://github.com/perfwatcher/collectm/blob/master/src/plugins/collectdCompat.js#L230 should be using `{}` instead of `[]` given how you access them as `results[interface_name].if_octets_tx`.

I did this simplification in `get_interface` to make sure that the bug wasn't being caused there. To me it looks simpler and less error prone. But in any case, please note that I'm not experienced with js and I trust that you'll chose the best approach. My main interest here is that it works as expected.

Regards,
Dimitris

**Comment by dimrozakis** [comment]
I opened a PR for node-collectdout: https://github.com/astro/node-collectdout/pull/7

**Comment by ymettier** [comment]
Hello,

Your updates/suggestions are included/merged into Collectm-1.5.1.
Thanks for your help.

Regards,
Yves
